app/services/post_service.rb

```diff
@@ -0,0 +1,65 @@
+class PostService
+  attr_reader :post
+
+  def initialize(params)
+    @id = params[:id]
+    @user = params[:user]
+    @oembed = params[:oembed]
+  end
+
+  def present_json
+    PostPresenter.new(post, user)
+  end
+
+  def present_interactions_json
+    PostInteractionPresenter.new(post, user)
+  end
+
+  def present_oembed
+    OEmbedPresenter.new(post, oembed)
+  end
+
+  def assign_post_and_mark_notifications
+    assign_post
+    mark_corresponding_notifications_read if user
+  end
+
+  def assign_post
+    if user
+      @post = Post.find_non_public_by_guid_or_id_with_user(id, user)
+    else
+      @post = Post.find_public(id)
+    end
+  end
```

Comment:
The service is hardly useful without this method called, in fact it will throw some ugly errors in some methods if it wasn't called prior. I think it's fair to do the fetching in the initializer already, if you're going to create it, you're going to use some method that requires fetching it too.